rfd/XXXX-upgrade-system.md

```diff
@@ -0,0 +1,1222 @@
+The second option will require a decent amount of specialized support and will be added later down the line. Script
+installers would likely need to be amended in some way to work correctly with a local rollback scheme. The
+details of how exactly local rollbacks should function are TBD. Some possibilities include:
+
+- Initially install new versions to a pending location (e.g. `/usr/local/bin/teleport.pending`).  Have teleport
+automatically fork a background monitor and `exec`s into the pending binary if it is detected on startup. If the
+background monitor observes that its requirements are met, it moves the pending binary to the active location,
+replacing the previous install.
+
+- Formally embrace the idea of multiple concurrently installed teleport versions
+and provide a thin "proxy binary" that can seamlessly `exec` into the current target version based on some filesystem
+config, potentially launching a background monitor of a different version first depending on said config. This has the
+downside of introducing a new binary, but the upside of eliminating the need for messy move/rename schemes.
+
+- Fully install the new version, creating a backup of the previous version first. Rely on an external mechanism for
+ensuring that the monitor/revert process gets run (e.g. by registering a new systemd unit). This has the upside of
+probably being compatible with script-based installers without any changes (teleport could create the backup and register
+the unit before starting the script), but has the downside of introducing an external dependency.
+
+
+## Implementation Plan
+
+There are a lot of individual components and features discussed in this RFD. As such,
+implementation will be divided into phases with multiple iterative releases consisting
+of subsets of the final feature set.
+
+
+### Inventory Status/Control Setup
+
+This phase sees no meaningful user-facing features added, but is the building block upon
+which most of the rest of the features are built.
+
+Instance-level status and control stream:
+
+- Refactor agent certificate logic to support advertising multiple system roles on a
+single cert (currently each service has its own disjoint certificate).
+
+- Implement per-instance bidirectional GRPC control stream capable of advertising all
+services running on a given teleport instance, and accepting commands directly from the
+controlling auth server.
+
+Improved inventory version tracking:
+
+- Improve teleport's self-knowledge so that instances can heartbeat detailed build
+attributes (arch, target os, fips status, etc).
+
+- Add new server inventory resource and `tctl inventory ls` command for viewing all
+instances w/ build info and services.
+
+With the above changes in place, we will have the ability to reliably inspect per-instance
+state regardless of running services, and each auth server will have a bidirectional handle
+to its connected nodes, allowing for real-time signaling.
+
+
+### Notification-Only System (?)
```

Comment:
@fspmarshall @xinding33 @russjones I would suggest we enable this notification-only system by default for all OSS or at least have it as an opt-out during install because it's in user's benefit. 

We should add UI notifications and tsh command outputs every time users use an obsolete system. 

For example, in UI and CLI:

```
$ tsh ls
WARNING: you are using obsolete Teleport that has CVE .... Upgrade the system immediately.
```

Can we make this system to be enabled by default for OSS and have opt-out for enterprise (more sensitive to outbound network calls). Can we make this a part of the Notification-only system deliverable?

I think this will deliver maximum security value with minimum engineering efforts.